File: ts/D3.ts
import { readFileSync } from 'fs';
import { timeAndPrint } from './time-and-print';

function parseInput(): string[] {
    const map = [];
    const content = readFileSync('../inputs/D3.input','utf8');
    for (let line of content.split("\r\n")) {
        map.push(line);
    }
    return map;
}

function treeSlope(map: string[], dr: number, dc: number): number {
    let count = 0;
    let r = 0, c = 0, w = map[0].length;
    while (r < map.length) {
        if (map[r][c % w] == '#')
            count++;
        r += dr;
        c += dc;
    }
    return count;
}

function p1() {
    const map = parseInput();
    return treeSlope(map, 1, 3);
}

function p2() {
    const map = parseInput();
    return treeSlope(map, 1, 1)
         * treeSlope(map, 1, 3)
         * treeSlope(map, 1, 5)
         * treeSlope(map, 1, 7)
         * treeSlope(map, 2, 1);
}

timeAndPrint(p1);
timeAndPrint(p2);

File: rs/src/d3.rs
use std::fs;

pub fn p1() -> usize {
   return tree_slope(&parse_input(), 1, 3);
}

pub fn p2() -> usize {
   let map = parse_input();
   return tree_slope(&map, 1, 1)
        * tree_slope(&map, 1, 3)
        * tree_slope(&map, 1, 5)
        * tree_slope(&map, 1, 7)
        * tree_slope(&map, 2, 1);
}

fn tree_slope(map: &Vec<Vec<char>>, dr: usize, dc: usize) -> usize {
    let mut r = 0;
    let mut c = 0;
    let mut trees = 0;
    let width = map[0].len();

    while r < map.len() {
        let current = map[r][c % width];

        if current == '#' {
            trees += 1;
        }

        r += dr;
        c += dc;
    }

    return trees;
}

fn parse_input() -> Vec<Vec<char>> {
    let contents = fs::read_to_string("../inputs/D3.input").unwrap();
    let mut map = Vec::new();

    for line in contents.split("\r\n") {
       map.push(line.chars().collect());
    }

    return map;
}